[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99580
* __->__ #99399
* #99398

This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

kimishpatel committed:
[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config

This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -28,6 +35,22 @@
+_QUANT_CONFIG_TO_ANNOTATOR = {}
+
+
+def register_annotator(quantization_configs: List[QuantizationConfig]):
```
why is this indexed by quantization_config? are we still planning to support module fqn/type based config or other existing features in QConfigMapping or do we just want to support global qconfig?

kimishpatel replied:
So the way I imagined this to be used is like this:
qnnpack, i will rename this to xnnpack, quantizer has two configs:
symmetric_static_config
symmetric_dynamic_config.

THis decorator registers annotation function for each.

Now if I want to quantize a model with dynamic q linear for linear and static quant for the rest, I would do it like this with composable quantizer.
```
dq_linear_config = QNNPACKQuantizer.get_dynamic_linear_config()
dq_linear_quantizer = QNNPACKQuantizer().set_global(dq_linear_config)

static_qconfig = QNNPACKQuantizer.get_symmetric_config()
static_quantizer = QNNPACKQuantizer().set_global(static_config)

composed_quantizer = ComposableQuantizer([dq_linear_quantizer, static_quantizer])
```

We can alternatively do this via:
```
qnnpack_quantizer = QNNPACKQuantizer().set_global(static_config)
# following api doesnt yet exist
qnnpack_quantizer.set_module_type_config(torch.nn.Linear, QNNPACKQuantizer.get_dynamic_linear_config()
```

I think both are fine. RIght now I am doing this mainly because it felt somewhat easier to use, but I am open to just doing via `set_module_config` or similar API.

kimishpatel replied:
Let me think a bit more. If it makes sense to just set_module_type_config api, then I will remove the config-to-annotator mapping part from the PR and follow up separately.

kimishpatel replied:
Ok I thought about this and I dont think we should do

```
dq_linear_config = QNNPACKQuantizer.get_dynamic_linear_config()
dq_linear_quantizer = QNNPACKQuantizer().set_global(dq_linear_config)

static_qconfig = QNNPACKQuantizer.get_symmetric_config()
static_quantizer = QNNPACKQuantizer().set_global(static_config)

composed_quantizer = ComposableQuantizer([dq_linear_quantizer, static_quantizer])

```

kimishpatel replied:
But for now I will use annotator api until we do this better

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

jerryzh168 commented:
just to confirm the plan is to follow up with set_module_type_config api right?

jerryzh168 reviewed (approved):
accepting to unblock for now, we can revisit the API a bit later

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2E][Quant] Refactor quantizer and qnnpack qantizer code to support dqlinear config"


This diff introduces a few refactors:

- Move observer creation to utils.py.
- Use quantization spec to supply args to observers.
- Use annotation function registration corresponding QuantizationConfig. This
  will be later used in dynamic quantized linear.

Differential Revision: [D45073790](https://our.internmc.facebook.com/intern/diff/D45073790/)

[ghstack-poisoned]

kimishpatel commented:
@pytorchbot merge
